### PR TITLE
Fix for SNAP-672, SNAP-394, SNAP-475, SNAP-353

### PR DESCRIPTION
**Pull request by ymahajan**
## Changes proposed in this pull request
Bug fixes in streaming (SNAP-672, SNAP-394, SNAP-475, SNAP-353)
## Patch testing
Added a new unit test
modified existing test 
## Other PRs 
No Other PR's required.

### CODE REVIEW

**Comment by ymahajan** [comment]
@rishitesh @hbhanawat  , please review. Would like to get these released with Sumedh's changes of stock spark

File: snappy-core/src/main/scala/org/apache/spark/sql/streaming/SchemaDStream.scala [hunk 1]
```diff
@@ -53,13 +53,25 @@ final class SchemaDStream(@transient val snsc: SnappyStreamingContext,
     super.filter(filterFunc)
   }
 
+ /* override def repartition(numPartitions: Int): SchemaDStream = {
```
Comment (rishitesh) [inline snappy-core/src/main/scala/org/apache/spark/sql/streaming/SchemaDStream.scala]: Remove commented code

File: snappy-dunits/src/test/scala/io/snappydata/dunit/streaming/TweetToRowsConverter.scala [hunk 2]
```diff
@@ -2,15 +2,15 @@ package io.snappydata.dunit.streaming
...
 import org.apache.spark.sql.streaming.StreamToRowsConverter
 import org.apache.spark.unsafe.types.UTF8String
 
 class TweetToRowsConverter extends StreamToRowsConverter with Serializable {
 
-  override def toRows(message: Any): Seq[InternalRow] = {
+  override def toRows(message: Any): Seq[Row] = {
     val status: Status = message.asInstanceOf[Status]
-    Seq(InternalRow.fromSeq(Seq(status.getId,
+    Seq(Row.fromSeq(Seq(status.getId,
```
Comment (rishitesh) [inline snappy-dunits/src/test/scala/io/snappydata/dunit/streaming/TweetToRowsConverter.scala]: Catalyst layer will convert a String to UtfString internally. Why we needed to do it?

File: snappy-core/src/main/scala/org/apache/spark/sql/streaming/DirectKafkaStreamSource.scala [hunk 3]
```diff
@@ -59,11 +59,13 @@ final class DirectKafkaStreamRelation(
...
-      val ckd: ClassTag[Decoder[Any]] = ClassTag(Utils.getContextOrSparkClassLoader.loadClass(KD))
-      val cvd: ClassTag[Decoder[Any]] = ClassTag(Utils.getContextOrSparkClassLoader.loadClass(VD))
-      KafkaUtils.createDirectStream[Any, Any, Decoder[Any], Decoder[Any]](context,
-        kafkaParams, topicsSet)(ck, cv, ckd, cvd).map(_._2).flatMap(rowConverter.toRows)
-    }
+    val ck: ClassTag[Any] = ClassTag(Utils.getContextOrSparkClassLoader.loadClass(K))
+    val cv: ClassTag[Any] = ClassTag(Utils.getContextOrSparkClassLoader.loadClass(V))
+    val ckd: ClassTag[Decoder[Any]] = ClassTag(Utils.getContextOrSparkClassLoader.loadClass(KD))
+    val cvd: ClassTag[Decoder[Any]] = ClassTag(Utils.getContextOrSparkClassLoader.loadClass(VD))
+    val converter = CatalystTypeConverters.createToCatalystConverter(schema)
```
Comment (rishitesh) [inline snappy-core/src/main/scala/org/apache/spark/sql/streaming/DirectKafkaStreamSource.scala]: This line is repetitive. Can we push it to some super class ?

File: snappy-core/src/main/scala/org/apache/spark/sql/streaming/StreamBaseRelation.scala [hunk 4]
```diff
@@ -79,6 +79,9 @@ abstract class StreamBaseRelation(options: Map[String, String])
   override val needConversion: Boolean = false
 
   override def buildScan(): RDD[Row] = {
+    if (context.getState() == StreamingContextState.STOPPED ){
```
Comment (rishitesh) [inline snappy-core/src/main/scala/org/apache/spark/sql/streaming/StreamBaseRelation.scala]: We need to discuss this more. If StreamingContext is stopped, how we are going to stop , clean-up other artefacts. 
One is catalog cleanup. What about already running queries. We should see how Spark jobs are handled on a stopped streaming context ( without stopping SparkContext)

File: snappy-tools/src/test/scala/io/snappydata/app/streaming/StreamingSuite.scala [hunk 5]
```diff
@@ -503,17 +550,17 @@ class TweetToRowsConverter extends StreamToRowsConverter with Serializable {
...
 
-  override def toRows(message: Any): Seq[InternalRow] = {
+  override def toRows(message: Any): Seq[Row] = {
     val status: Status = TwitterObjectFactory.createStatus(message.asInstanceOf[String])
     TwitterObjectFactory.getRawJSON(message)
-    Seq(InternalRow.fromSeq(Seq(status.getId,
+    Seq(Row.fromSeq(Seq(status.getId,
       UTF8String.fromString(status.getText),
       UTF8String.fromString(status.getUser().getName),
       UTF8String.fromString(status.getUser.getLang),
```
Comment (rishitesh) [inline snappy-tools/src/test/scala/io/snappydata/app/streaming/StreamingSuite.scala]: rest all looks allright to me.

File: snappy-core/src/main/scala/org/apache/spark/sql/streaming/FileStreamSource.scala [hunk 6]
```diff
@@ -62,6 +61,9 @@ final class FileStreamRelation(
   val directory = options(DIRECTORY)
 
   // TODO: Yogesh, add support for other types of files streams
-  override protected def createRowStream(): DStream[InternalRow] =
+  override protected def createRowStream(): DStream[InternalRow] = {
+    val converter = CatalystTypeConverters.createToCatalystConverter(schema)
     context.textFileStream(directory).flatMap(rowConverter.toRows)
+        .map(converter(_).asInstanceOf[InternalRow])
+  }
 }
```
Comment (sumwale) [inline snappy-core/src/main/scala/org/apache/spark/sql/streaming/FileStreamSource.scala]: @ymahajan perhaps we can have something similar BaseRelation.needConversion for some possible optimization in this. If the implementation says needConversion is false, then assume the converter returns InternalRows and treat it likewise (i.e. returning a Seq[InternalRow] even though the signature says Seq[Row]).

**Pull request merged**